src/pykeen/training/training_loop.py

```diff
@@ -181,22 +215,69 @@ def train(
         # In some cases, e.g. using Optuna for HPO, the cuda cache from a previous run is not cleared
         torch.cuda.empty_cache()
 
-        result = self._train(
-            num_epochs=num_epochs,
-            batch_size=batch_size,
-            slice_size=slice_size,
-            label_smoothing=label_smoothing,
-            sampler=sampler,
-            continue_training=continue_training,
-            only_size_probing=only_size_probing,
-            use_tqdm=use_tqdm,
-            use_tqdm_batch=use_tqdm_batch,
-            tqdm_kwargs=tqdm_kwargs,
-            stopper=stopper,
-            result_tracker=result_tracker,
-            sub_batch_size=sub_batch_size,
-            num_workers=num_workers,
-        )
+        # A checkpoint root is always created to ensure a fallback checkpoint can be saved
+        if checkpoint_root is None:
+            checkpoint_root = PYKEEN_DEFAULT_CHECKPOINT_DIR
+        checkpoint_root = pathlib.Path(checkpoint_root)
+        checkpoint_root.mkdir(parents=True, exist_ok=True)
+        logger.debug('using checkpoint_root at %s', checkpoint_root)
+
+        # If a checkpoint file is given, it must be loaded if it exists already
+        save_checkpoints = False
+        if checkpoint_file:
```

Comment:
There's an awful lot of branching logic here. are there unit tests for all of them?

Reply:
Imo unit tests wouldn't make sense here, as all this branching logic simply is to inform the user about what is happening when using checkpoints combined with the consequences of different choices or possible misconfigurations. Or do you have something else in mind?

Reply:
I think it's okay for now. After a few rounds of refactoring, it's much more obvious